components/core/src/clp_s/ArchiveReader.cpp

```diff
@@ -3,81 +3,187 @@
     }
 
-    std::set<int32_t> schema_ids;
-    boost::filesystem::directory_iterator iter(encoded_messages_dir);
-    boost::filesystem::directory_iterator end;
-
-    // Get all schema ids
-    for (; iter != end; ++iter) {
-        if (boost::filesystem::is_regular_file(iter->path())) {
-            std::string path = iter->path().rbegin()->string();
-            if (false == path.empty() && std::all_of(path.begin(), path.end(), ::isdigit)) {
-                schema_ids.insert(std::stoi(path));
-            }
+    for (size_t i = 0; i < num_schemas; i++) {
+        int32_t schema_id;
+        uint64_t num_messages;
+        size_t table_offset;
+
+        if (auto error = m_metadata_decompressor.try_read_numeric_value(schema_id);
+            ErrorCodeSuccess != error)
+        {
+            throw OperationFailed(error, __FILENAME__, __LINE__);
+        }
+
+        if (auto error = m_metadata_decompressor.try_read_numeric_value(num_messages);
+            ErrorCodeSuccess != error)
+        {
+            throw OperationFailed(error, __FILENAME__, __LINE__);
+        }
+
+        if (auto error = m_metadata_decompressor.try_read_numeric_value(table_offset);
+            ErrorCodeSuccess != error)
+        {
+            throw OperationFailed(error, __FILENAME__, __LINE__);
         }
+
+        m_id_to_table_metadata[schema_id] = {num_messages, table_offset};
+        m_schema_ids.push_back(schema_id);
     }
+    m_metadata_decompressor.close();
+}
+
+void ArchiveReader::read_dictionaries_and_metadata() {
+    read_variable_dictionary();
+    read_log_type_dictionary();
+    read_array_dictionary();
+    read_metadata();
+}
 
-    if (schema_ids.empty()) {
+std::unique_ptr<SchemaReader>
+ArchiveReader::read_table(int32_t schema_id, bool should_extract_timestamp) {
+    constexpr size_t cDecompressorFileReadBufferCapacity = 64 * 1024;  // 16 KB
+
+    if (m_id_to_table_metadata.count(schema_id) == 0) {
         throw OperationFailed(ErrorCodeFileNotFound, __FILENAME__, __LINE__);
     }
 
-    // Open schema readers and load encoded messages
-    for (int32_t schema_id : schema_ids) {
-        auto& schema = m_schema_map[schema_id];
-        auto schema_reader = new SchemaReader(m_schema_tree, schema_id);
-        schema_reader->open(encoded_messages_dir + "/" + std::to_string(schema_id));
-
-        ReaderUtils::append_reader_columns(
-                schema_reader,
-                schema,
-                m_schema_tree,
-                m_var_dict,
-                m_log_dict,
-                m_array_dict,
-                m_timestamp_dict,
-                false
-        );
-
-        schema_reader->load();
-        m_schema_id_to_reader[schema_id] = schema_reader;
+    auto schema_reader = std::make_unique<SchemaReader>(
+            m_schema_tree,
+            schema_id,
+            m_id_to_table_metadata[schema_id].num_messages
+    );
+    append_reader_columns(schema_reader, should_extract_timestamp);
```

Comment:
It's sort of weird reading this code since it looks like we're creating a unique_ptr, handing it off to this append_reader_columns function, and also returning it a few lines later.

It might be worth at least adding a comment that the schema_reader unique_ptr is being passed by reference here.

Reply:
Probably it's better to create a unique_ptr inside append_reader_columns and return it?

Reply:
Yeah, that's a good idea. Maybe also rename it to `create_schema_reader` or something like that since it'll be doing all of the initialization.